lib/posix.lua.in

```diff
@@ -519,27 +519,42 @@ if _DEBUG ~= false then
   end
 end
 
-local function prepare_args (pat_mark)
-  local pat, mark = "*", 0
-  if type (pat_mark) == "table" then
-    if type (pat_mark.pattern) == "string" then
-      pat = pat_mark.pattern
-    end
-    if pat_mark.MARK then
-      mark = M.GLOB_MARK
-    end
-  elseif type (pat_mark) == "string" then
-    pat = pat_mark
-  end
-
-  return pat, mark
+--- Find all files in this directory matching a shell pattern.
+-- The flag MARK appends a trailing slash to matches that are directories.
+-- @function glob
+-- @param[opt={}] string|table args either a glob pattern string or a table
+--    with optional keys `pattern` and `MARK`. If no argument is given, the
+--    function uses a default pattern of "*" and MARK is false.
+-- @treturn A table of zero or more matches to the pattern.
+local function glob (args)
+  local posix_glob = require "posix.glob"
+  -- Support previous `glob ".*"` style calls.
+  if type(args) == "string" then
+    args = {pattern=args}
+  elseif type(args) ~= "table" then
+    args = {}
+  end
+  local flags = 0
+  if args.MARK then
+    flags = posix_glob.GLOB_MARK
+  end
+  return posix_glob.glob(args.pattern, flags)
 end
 
-local cglob = M.glob
-
-M.glob = function (...)
-  local pat, mark = prepare_args(...)
-  return cglob(pat, mark)
+if _DEBUG ~= false then
+  M.glob = function (...)
+    local argt = {...}
+    local argtype = type(argt[1])
+    local validtypes = { ["table"] = true, ["string"] = true, ["nil"] = true }
```

Comment:
same here with scoping... instead of constructing a new validtypes table every time we call, move it into file local scope just after the `if _DEBUG ~= false then` line.